packages/common/http/src/transfer_cache.ts

```diff
@@ -61,24 +68,56 @@ export function transferCacheInterceptorFn(
         break;
     }
 
+    // We want to warn users accessing a header provided from the cache
+    // That HttpTransferCache alters the headers
+    // The warning will be logged a single time by HttpHeaders instance
+    let shownUsageWarning = false;
+    const headers = typeof ngDevMode === 'undefined' || ngDevMode ?
```

Comment:
We can move this logic (the whole block) to a separate function, so that we keep this function more readable. Note: we should move it to a separate standalone function (not a method on a class).

```suggestion
    let headers = new HttpHeaders(response.headers);
    if (typeof ngDevMode === 'undefined' || ngDevMode) {
      // Append extra logic in dev mode to produce a warning when a header
      // that was not transferred to the client is accessed in the code via `get`
      // and `has` calls.
      headers = appendMissingHeadersDetection(headers, headersToInclude);
    }
```

We can also calculate `headersToInclude` ahead of this call and pass it as an argument. In this case, we can be more precise and only output the warning when a header outside of the `headersToInclude` list is accessed. That would also mean that we can't intercept `keys` and `getAll`, only `get` and `has`, which should be acceptable.